GVFS/GVFS.Virtualization/FileSystem/FileSystemVirtualizer.cs

```diff
@@ -269,18 +269,15 @@ protected void OnHardLinkCreated(string relativeExistingFilePath, string relativ
         {
             try
             {
-                if (!string.IsNullOrEmpty(relativeNewLinkPath))
-                {
-                    bool pathInDotGit = FileSystemCallbacks.IsPathInsideDotGit(relativeNewLinkPath);
+                bool newLinkPathInDotGit = FileSystemCallbacks.IsPathInsideDotGit(relativeNewLinkPath);
 
-                    if (pathInDotGit)
-                    {
-                        this.OnDotGitFileOrFolderChanged(relativeNewLinkPath);
-                    }
-                    else
-                    {
-                        this.FileSystemCallbacks.OnFileHardLinkCreated(relativeNewLinkPath);
-                    }
+                if (newLinkPathInDotGit)
+                {
+                    this.OnDotGitFileOrFolderChanged(relativeNewLinkPath);
+                }
+                else
+                {
+                     this.FileSystemCallbacks.OnFileHardLinkCreated(relativeNewLinkPath, relativeExistingFilePath);

```

Comment:
It looks like with this change `this.FileSystemCallbacks.OnFileHardLinkCreated` would still be called in this scenario (when it should not be):

- `relativeNewLinkPath == string.Empty`
- `relativeExistingFilePath` -> Path inside the .git folder

And `FileSystemCallbacks.OnFileHardLinkCreated` is not getting called in this scenario (when it should be):

- `relativeNewLinkPath` -> Path in the .git folder
- `relativeExistingFilePath` -> Path inside the working tree

I think you should be able to write a functional test that catches the second scenario (I suspect the first scenario is only a perf issue)

Reply:
Let me know what you think now.  I can add a specific functional test for relativeNewLinkPath from .git, and relativeExistingFile path from the working tree, but I'm a bit cautious of adding too many functional tests (already added a bunch).  I would expect this to be a rare case as well.  Let me know what you think.  I can add the test if we really think we need it.

Reply:
With the current check this scenario will not be handled correctly:

`relativeNewLinkPath` -> .git\\git_file
`relativeExistingFilePath` -> ""

Because it will result in:

`if (!(false && false))`

And `OnFileHardLinkCreated` will be called when neither path is in the working tree.

What would you think of making this check a bit lengthier, but perhaps easier to understand:

```
// call OnFileHardLinkCreated if either relativeNewLinkPath or relativeExistingFilePath are in the working tree
if ( (!newLinkPathInDotGit && !string.IsNullOrEmpty(relativeNewLinkPath) ||
     (!relativePathInDotGit && !string.IsNullOrEmpty(relativeExistingFilePath) )
```

Or even the *slightly* less efficient:

```
bool newLinkInWorkingDir = !newLinkPathInDotGit && !string.IsNullOrEmpty(relativeNewLinkPath);
bool existingFileInWorkingDir = !relativePathInDotGit && !string.IsNullOrEmpty(relativeExistingFilePath);
if ( newLinkInWorkingDir || existingFileInWorkingDir)
```

I think it would be find to always compute `newLinkInWorkingDir` and `existingFileInWorkingDir` to make the condition easier to read.

Reply:
I think the code is actually works the way it stands.  

The second scenario was indeed be false, but 
```
if (newLinkPathInDotGit)
{
   this.OnDotGitFileOrFolderChanged(relativeNewLinkPath);
}
```
Would have been called previously making the status cache invalid.  
Calling OnFileHardLinkCreated does nothing for the scenario you called out as far as I can see since the path is a in a git folder.  Please let me know if I'm incorrect. :-)

That said the way the code stands now is obviously confusing so i'm updating it an a separate commit to be more similar with how FileRename is handled.  Since that one does the "IsNullOrEmpty" check in the FileSystemCallback I think it's fine to do the same here.  I"ll wait until your back to give a thumbs up and confirm this new change isn't confusing.